### PR TITLE
[JENKINS-33273] Allow lightweight checkouts to be used from CpsScmFlowDefinition

### PR DESCRIPTION
**Pull request by jglick**
Analogue of https://github.com/jenkinsci/workflow-multibranch-plugin/pull/47 for standalone Pipeline projects.

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -199,7 +199,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>git</artifactId>
-            <version>2.4.1</version>
+            <version>3.0.2-beta-2</version>
```
Comment (jglick) [inline pom.xml]: Will pick up https://github.com/jenkinsci/git-plugin/pull/467 when available.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java [hunk 2]
```diff
@@ -73,18 +76,36 @@ public String getScriptPath() {
...
             }
         }
-        FilePath dir;
         Queue.Executable _build = owner.getExecutable();
         if (!(_build instanceof Run)) {
             throw new IOException("can only check out SCM into a Run");
         }
         Run<?,?> build = (Run<?,?>) _build;
+        if (lightweight) {
+            SCMFileSystem fs = SCMFileSystem.of(build.getParent(), scm);
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java]: use try with resources

**Comment by jglick** [comment]
@reviewbybees done

(but not safe to merge until @stephenc has sorted out the dependencies)

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/help-lightweight.html [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<div>
+    If selected, try to obtain the Pipeline script contents directly from the SCM without performing a full checkout.
+    The advantage of this mode is its efficiency; however, you will not get any changelogs or polling based on the SCM.
```
Comment (stephenc) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/help-lightweight.html]: You should be able to get the changelogs with the `SCMFilesystem` API...

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly [hunk 4]
```diff
@@ -29,6 +29,9 @@ THE SOFTWARE.
     <f:entry field="scriptPath" title="${%Script Path}">
         <f:textbox default="Jenkinsfile"/>
     </f:entry>
+    <f:entry field="lightweight" title="${%Lightweight checkout}">
```
Comment (stephenc) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly]: I would think you would at most want an advanced option to *not* try lightweight. If you can do lightweight, why wouldn't you want to?

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/help-lightweight.html (hunk 3 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/help-lightweight.html]: Would require a significant rewrite. Currently `workflow-job` knows to record changelogs, or consider polling, based on `SCMListener` methods which cannot be called in lightweight mode. I considered such a large change (APIs etc.) out of scope here, especially given that

* many people will not care about changelogs or polling for the SCM containing the script
* if you `checkout scm` the same repo later, you will get those anyway

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly (hunk 4 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly]: I preferred to err on the side of leaving existing behavior alone by default for now.

* lack of changelog/polling support (see below)
* new code path with likely bugs compared to original

File: pom.xml [hunk 5]
```diff
@@ -95,7 +95,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>scm-api</artifactId>
-            <version>1.1</version>
+            <version>2.0.1</version>
```
Comment (jglick) [inline pom.xml]: Should ideally be bumped up to whatever release actually makes it to the UC.

**Comment by jglick** [comment]
@reviewbybees done

Though I am still considering introducing an option or at least a flag to disable https://github.com/jenkinsci/workflow-multibranch-plugin/pull/47.

**Comment by recampbell** [comment]
This looks mergable?

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly (hunk 4 above)
Comment (HRMPW) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly]: This box is selected by default. I think it is safer to maintain existing behavior for existing Pipelines.

**Comment by jglick** [comment]
Needs a code review, and would like signoff from @HRMPW as well.

**Comment by HRMPW** [comment]
The UI change is minimal and looks good. :bee:

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly (hunk 4 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition/config.jelly]: That is what we are doing—`default` is used for _new_ definitions.

**Pull request merged**